### PR TITLE
SALTO-2577 - Handle Categories order in brand

### PR DESCRIPTION
**Pull request by idoser96**
Added 'categories' field inside brands that have Guide activated
this field is initialized in fetch and used to update categories positions in deploy

---

The same filter will handle the categories and the brand's deploy together

---
_Release Notes_: 

Zendesk adapter:
* on zendesk fetch with brands with Guide, a new field will be added to brand, this field describes the order of the categories

---
_User Notifications_: 
Zendesk adapter:
* Brand may have 'categories' field inside
* Category's 'position' field is now hidden

### CODE REVIEW

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    )
+
+    const [onlyOrderChanges, mixedChanges, onlyNonOrderChanges] = sortBrandChanges(brandChanges)
+
+    const fieldsToIgnore: Set<string> = new Set<string>()
+    const orderChangesToApply: Change<InstanceElement>[] = []
+
+    for (const brandChange of [...onlyOrderChanges, ...mixedChanges]) {
+      const brandValue = brandChange.data.after.value
+      const categories = brandValue.categories.map((c: ReferenceExpression) => c.value)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:120]: we shouldn't assume all of these are reference expressions - some may have been omitted for various reasons (e.g. conflicting element ids / limited fetch scope)
please check these and either fail the operation with a relevant error message, or decide on handling for it

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+      changes,
+      change => getChangeData(change).elemID.typeName === BRAND_TYPE_NAME,
+    )
+
+    const [onlyOrderChanges, mixedChanges, onlyNonOrderChanges] = sortBrandChanges(brandChanges)
+
+    const fieldsToIgnore: Set<string> = new Set<string>()
+    const orderChangesToApply: Change<InstanceElement>[] = []
+
+    for (const brandChange of [...onlyOrderChanges, ...mixedChanges]) {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:118]: nit: there's a preference in the OSS code to avoid `for` loops and prefer functional style, can we use `.forEach()` or `.map()`?

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      const brandValue = brandChange.data.after.value
+      const categories = brandValue.categories.map((c: ReferenceExpression) => c.value)
+
+      categories.forEach((category: InstanceElement, i : number) => {
+        // Create a 'fake' change of the category's position
+        const beforeCategory = category.clone()
+        category.value.position = i
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:125]: do we want anything here besides the category's id and position? might be easier to create a change with just these two values instead of adding everything else to the ignore list? also see @edenhassid 's [implementation](https://github.com/salto-io/salto/blob/main/packages/zendesk-adapter/src/filters/help_center_parent_to_section.ts#L74-L93) for making another API call to update a section's parent, that we can probably reuse here (maybe even share the code?)

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import _ from 'lodash'
+import { FilterCreator } from '../filter'
+import { BRAND_TYPE_NAME } from '../constants'
+import { deployChange, deployChanges } from '../deployment'
+import { LOGO_FIELD } from './brand_logo'
+
+export const CATEGORIES_FIELD = 'categories'
+
+// Lowest position index first, if there is a tie - the newer is first
+const compareCategoriesPosition = (a: InstanceElement, b: InstanceElement) : number => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:31]: I think it might be easier to use `_.sortBy()` for these - you can use `[x.value.position, -x.value.createdAt]` as the keys IIUC (and then we can probably have this implemented even inline?)

File: packages/zendesk-adapter/test/filters/categories_order.test.ts [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    beforeEach(async () => {
+      jest.clearAllMocks()
+      filter = filterCreator(createFilterCreatorParams({})) as FilterType
+    })
+
+    describe('on fetch', () => {
+      it('with Guide active', async () => {
+        // Should create categories order field
+        const brandWithGuide = createBrandInstance()
+        const categories = [createCategory(), createCategory()]
```
Comment (netama) [inline packages/zendesk-adapter/test/filters/categories_order.test.ts:83]: I think it's important to check the various edge cases for the overlaps etc here - let's add some more complex scenarios?

**Review by netama — COMMENTED** [review]
did a first pass - looks good overall, added some suggestions :)

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+  /** Insert the brand's categories into a field in it */
+  onFetch: async (elements: Element[]) => {
+    const categories = elements.filter(isInstanceElement).filter(e => e.elemID.typeName === 'category')
+    const brands = elements.filter(isInstanceElement)
+      .filter(e => e.elemID.typeName === BRAND_TYPE_NAME)
+
+    // Insert the categories of that brand to the brand, sorted
+    brands.forEach(brand => {
+      // If the brand doesn't have Guide activated, do nothing
+      if (!brand.value.has_help_center) { return }
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:96]: I actually prefer my way a lot more, keeps the code shorter and more readable (and there is no logic for just 'return')
wdyt?

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 3 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:125]: Will change it in this branch and will create a generic function on the next PR

File: packages/zendesk-adapter/test/filters/categories_order.test.ts (hunk 5 above)
Comment (idoser96) [reply packages/zendesk-adapter/test/filters/categories_order.test.ts:83]: Can you give me an example?

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 1 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:120]: How can it not be reference if we this is a custom field we put in 'fetch'?

File: packages/zendesk-adapter/src/config.ts [hunk 7]
```diff
@@ -1928,7 +1928,7 @@ export const DEFAULT_TYPES: ZendeskApiConfig['types'] = {
       standaloneFields: [{ fieldName: 'translations' }],
       sourceTypeName: 'categories__categories',
       fieldsToHide: FIELDS_TO_HIDE.concat(
-        { fieldName: 'id', fieldType: 'number' },
+        { fieldName: 'id', fieldType: 'number' }, { fieldName: 'position', fieldType: 'number' },
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/config.ts:1931]: Why list here? its a number

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+/* Split the changes into 3 groups:
+  onlyOrderChanges    - Brands with only categories order changes
+  mixedChanges        - Brands with categories order and other changes
+  onlyNonOrderChanges - Brands without any categories order changes
+ */
+const sortBrandChanges = (changes: Change<InstanceElement>[]) :
+    [ ModificationChange<InstanceElement>[],
+      ModificationChange<InstanceElement>[],
+      Change<InstanceElement>[]] => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:38]: 1. can we return an object with the types? will make this more readable IMO and make it less likely to have mistakes in the future when editing the function / return values
2. does something external rely on the three categories? I'm wondering if it might be cleaner to just return the list of changes that have order changes and the list of changes that have non-order changes even if they have overlaps?

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  const onlyOrderChanges : ModificationChange<InstanceElement>[] = []
+  const mixedChanges : ModificationChange<InstanceElement>[] = []
+  const onlyNonOrderChanges : Change<InstanceElement>[] = []
+
+  changes.forEach(change => {
+    if (isRemovalChange(change)) {
+      onlyNonOrderChanges.push(change)
+      return
+    }
+    // currently isn't supported because categories can't exist before brand
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:48]: wdym? they can be created in the same deploy, no?

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    if (isAdditionChange(change)) {
+      onlyNonOrderChanges.push(change)
+      return
+    }
+    const brandChanges = detailedCompare(change.data.before, change.data.after)
+    const hasAnyOrderChanges = brandChanges.some(c =>
+      c.id.getFullNameParts().includes(CATEGORIES_FIELD))
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:55]: we should be more careful here - please check this under the exact path (e.g. use `.createTopLevelParentID()` and make sure the `path` matches exactly, or compare to the exact expected elem id based on the change id).

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    const hasOnlyOrderChanges = brandChanges.every(c =>
+      c.id.getFullNameParts().includes(CATEGORIES_FIELD))
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:57]: same. in general we shouldn't use `.getFullNameParts()` for comparison, it's better to compare the structured elem id to avoid mistakes

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+  return [onlyOrderChanges, mixedChanges, onlyNonOrderChanges]
+}
+
+/**
+ * Handle everything related to brands
+ */
+const filterCreator: FilterCreator = ({ client, config }) => ({
+  /** Insert the brand's categories into a field in it */
+  onFetch: async (elements: Element[]) => {
+    const categories = elements.filter(isInstanceElement)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:77]: nit: we can filter all instances by `isInstanceElement` once and use them for subseequent per-type filtering

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:85]: we don't have this style anywhere in the code now, and I'm not sure I agree about readability :) it's just a bit too close to styles that can actually cause bugs (e.g. not having the surrounding brackets), and we definitely shouldn't have this style when there is a return value - so IMO it's better to avoid it. but we can discuss more if you want :)

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+    // Insert the categories of that brand to the brand, sorted
+    brands.forEach(brand => {
+      // If the brand doesn't have Guide activated, do nothing
+      if (!brand.value.has_help_center) { return }
+
+      // Lowest position index first, if there is a tie - the newer is first
+      const brandsCategories = _.sortBy(
+        categories.filter(c => c.value.brand === brand.value.id),
+        (c => [c.value.position, -c.value.created_at])
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:90]: just making sure, is `created_at` a string/number that can be compared as-is? or do we need to convert it to some timestamp?

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+      )
+
+      brand.value[CATEGORIES_FIELD] = brandsCategories
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:93]: let's make sure that even if we have two categories with the same elem id (which means they'll be dropped from the workspace), this behaves correctly

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+    const [brandChanges, leftoverChanges] = _.partition(
+      changes,
+      change => getChangeData(change).elemID.typeName === BRAND_TYPE_NAME,
+    )
+
+    const [onlyOrderChanges, mixedChanges, onlyNonOrderChanges] = sortBrandChanges(brandChanges)
+
+    const orderChangesToApply: Change<InstanceElement>[] = []
+
+    const orderChanges = [...onlyOrderChanges, ...mixedChanges]
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:108]: continuing from the comment above - if you are always merging these together, I think it will be cleaner if the helper function just returns the result you need. it seems like we are doing extra work to separate and merge these that I'm not sure is helping (but maybe I'm missing something)

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 1 above)
Comment (netama) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:111]: commented above on this one (before I saw this reply 🤦‍♀️ ) - it seems our problem is starting earlier then :) but we shouldn't figure out how to manage instances that may have been omitted, or at least have an error about them
also - we should have a change validator that ensures all instances of the relevant types appear in the relevant `order` field (we have these in the support instances, here it might require some tweaks)

File: packages/zendesk-adapter/test/filters/categories_order.test.ts [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+        const categories = [createCategory(), createCategory(), createCategory(), createCategory()]
+        categories[0].value.position = 0
+        categories[1].value.position = 0
+        categories[2].value.position = 1
+        categories[3].value.position = 1
+
+        categories[0].value.created_at = '0'
+        categories[1].value.created_at = '1'
+        categories[2].value.created_at = '1'
+        categories[3].value.created_at = '0'
```
Comment (netama) [inline packages/zendesk-adapter/test/filters/categories_order.test.ts:94]: if we already have a function, why not use it to set both the `id` and `created_at` correctly up front?

Comment (netama) [reply packages/zendesk-adapter/test/filters/categories_order.test.ts:94]: also, can we use realistic `created_at` values here please? we should make sure string comparison is really valid

File: packages/zendesk-adapter/test/filters/categories_order.test.ts [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+
+      const beforeFirstCategory = firstCategory.clone()
+      const beforeSecondCategory = secondCategory.clone()
+      const afterFirstCategory = firstCategory.clone()
+      const afterSecondCategory = secondCategory.clone()
+
+      beforeFirstCategory.value.position = 0
+      beforeSecondCategory.value.position = 1
+      afterFirstCategory.value.position = 1
+      afterSecondCategory.value.position = 0
```
Comment (netama) [inline packages/zendesk-adapter/test/filters/categories_order.test.ts:127]: same re using the helper function to create everything

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 15 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:108]: You're right, originally I didn't use onlyOrderChange but now i do

File: packages/zendesk-adapter/src/config.ts [hunk 18]
```diff
@@ -491,6 +491,7 @@ export const DEFAULT_TYPES: ZendeskApiConfig['types'] = {
       fieldTypeOverrides: [
         { fieldName: 'help_center_state', fieldType: 'string', restrictions: { enforce_value: true, values: ['enabled', 'disabled', 'restricted'] } },
         { fieldName: 'id', fieldType: 'number' },
+        { fieldName: 'categories', fieldType: 'list<unkown>' },
```
Comment (Eldardar) [inline packages/zendesk-adapter/src/config.ts:494]:
```suggestion
        { fieldName: 'categories', fieldType: 'list<unknown>' },
```

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 14 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:89]: Dropping the elements can cause confusions
its probably better to either drop everything or drop nothing

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 12 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/brands_filter.ts:71]: Do you mean first do something like
`const instanceElements = elements.filter(...)?`
and then use it?

File: packages/zendesk-adapter/src/filters/brands_filter.ts (hunk 8 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/brands_filter.ts]: Lowered the return to two types
I don't think return type should become a type, in my opinion its complicating reading and debugging the code, especially now when there are only two return values

File: packages/zendesk-adapter/src/filters/brands_filter.ts [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+    const { withOrderChanges, onlyNonOrderChanges } = sortBrandChanges(brandChanges)
+
+    const orderChangesToApply: Change<InstanceElement>[] = []
+    const orderChangeErrors: Error[] = []
+
+    withOrderChanges.filter(isAdditionOrModificationChange).forEach(brandChange => {
+      const brandValue = brandChange.data.after.value
+
+      if (brandValue.categories.some((c: Value) => !(c instanceof ReferenceExpression))) {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/brands_filter.ts:110]: please use `isReferenceExpression` - then this can be
```suggestion
      if (!brandValue.categories.every(isReferenceExpression)) {
```

**Review by netama — APPROVED** [review]
LGTM 🎈  with the follow-ups we discussed (sorry I missed the limitation earlier)
